vmm/src/acpi.rs

```diff
@@ -105,23 +125,120 @@ pub fn create_dsdt_table(
+    // Secure EL1 Timer Flags
+    gtdt.write(52, irqflags);
+    // Non-Secure EL1 Timer GSIV
+    gtdt.write(56, (ARCH_TIMER_NS_EL1_IRQ + 16) as u32);
+    // Non-Secure EL1 Timer Flags
+    gtdt.write(60, (irqflags | ACPI_GTDT_CAP_ALWAYS_ON) as u32);
+    // Virtual EL1 Timer GSIV
+    gtdt.write(64, (ARCH_TIMER_VIRT_IRQ + 16) as u32);
+    // Virtual EL1 Timer Flags
+    gtdt.write(68, irqflags);
+    // EL2 Timer GSIV
+    gtdt.write(72, (ARCH_TIMER_NS_EL2_IRQ + 16) as u32);
+    // EL2 Timer Flags
+    gtdt.write(76, irqflags);
+
+    gtdt.update_checksum();
+
+    gtdt
+}
+
+#[cfg(target_arch = "aarch64")]
+fn create_spcr_table(base_address: u64, gsi: u32) -> SDT {
+    // SPCR
+    let mut spcr = SDT::new(*b"SPCR", 80, 2, *b"CLOUDH", *b"CHSPCR  ", 1);
+    // Size in bits of given register
+    spcr.write(41, 8u8);
+    // ACPI 3.0: Minimum Access size (ACPI 3.0)
+    spcr.write(43, 1u8);
+    // 64-bit address
+    spcr.write(44, (base_address as u64).to_le());
+    // Interrupt Type: Bit[3] ARMH GIC interrupt
+    spcr.write(48, (1 << 3) as u8);
+    // GSIV used by the UART
+    spcr.write(54, (gsi as u32).to_le());
+    // Baud Rate: 3 = 9600
+    spcr.write(58, 3u8);
+    // Stop Bits: 1 Stop bit
+    spcr.write(60, 1u8);
+    // Flow Control: Bit[1] = RTS/CTS hardware flow control
+    spcr.write(61, (1 << 1) as u8);
+    // PCI Device ID: Not a PCI device
+    spcr.write(64, 0xffff_u16);
+    // PCI Vendor ID: Not a PCI device
+    spcr.write(66, 0xffff_u16);
+
+    spcr.update_checksum();
+
+    spcr
+}
+
+#[cfg(target_arch = "aarch64")]
+fn create_iort_table() -> SDT {
+    const ACPI_IORT_NODE_ITS_GROUP: u8 = 0x00;
+    const ACPI_IORT_NODE_PCI_ROOT_COMPLEX: u8 = 0x02;
+
+    // IORT
+    let mut iort = SDT::new(*b"IORT", 124, 2, *b"CLOUDH", *b"CHIORT  ", 1);
+    // Nodes: PCI Root Complex, ITS
+    // Note: We currently do not support SMMU
+    iort.write(36, (2u32).to_le());
+    iort.write(40, (48u32).to_le());
+
+    // ITS group node
+    iort.write(48, ACPI_IORT_NODE_ITS_GROUP as u8);
+    // Length of the ITS group node in bytes
+    iort.write(49, (24u16).to_le());
+    // ITS counts
+    iort.write(64, (1u32).to_le());
+
+    // Root Complex Node
+    iort.write(72, ACPI_IORT_NODE_PCI_ROOT_COMPLEX as u8);
+    // Length of the root complex node in bytes
+    iort.write(73, (52u16).to_le());
+    // Mapping counts
+    iort.write(80, (1u32).to_le());
+    // Offset from the start of the RC node to the start of its Array of ID mappings
+    iort.write(84, (32u32).to_le());
+    // Fully coherent device
+    iort.write(88, (1u32).to_le());
+    // CCA = CPM = DCAS = 1
+    iort.write(95, 3u8);
+    // Identity RID mapping covering the whole input RID range
+    iort.write(108, (0xffff_u32).to_le());
+    // id_mapping_array_output_reference should be
+    // the ITS group node (the first node) if no SMMU
+    iort.write(116, (48u32).to_le());
+
+    iort.update_checksum();
+
+    iort
+}
+
 pub fn create_acpi_tables(
```

Comment:
Rather than all the conditional code in this function, maybe we could split it into separate versions for aarch64 and x86_64 and extract common code (like FADT.)

Reply:
Please look at the latest 2 commits. The code of this file was largely refactored.